### ISSUE TITLE
Tilt Sensor Init Sequence repeated twice

### ISSUE THREAD

**Issue by marco-bertani** [ysard/MyOwnBricks#1]
Hi, 

I was checking out the file `TiltSensor.cpp` trying to decode the messages just for fun and completion.
This is my result:

````
\x49\x03\x02\xB7
\x52\x00\xC2\x01\x00\x6E
\x5F\x00\x00\x00\x10\x00\x00\x00\x10\xA0

\x9B\x00\x4C\x50\x46\x32\x2D\x43\x41\x4C\x6F
\x9B\x01\x00\x00\x34\xC2\x00\x00\x34\x42\xE5
\x9B\x02\x00\x00\xC8\xC2\x00\x00\xC8\x42\xE6
\x9B\x03\x00\x00\x34\xC2\x00\x00\x34\x42\xE7
\x93\x04\x43\x41\x4C\x00\x26
\x8B\x05\x10\x00\x61
\x93\x80\x03\x00\x03\x00\xEC

\xA2\x00\x4C\x50\x46\x32\x2D\x43\x52\x41\x53\x48\x00\x00\x00\x00\x00\x00\x53 <- mode 2

\x9A\x01\x00\x00\x00\x00\x00\x00\xC8\x42\xEE
\x9A\x02\x00\x00\x00\x00\x00\x00\xC8\x42\xED
\x9A\x03\x00\x00\x00\x00\x00\x00\xC8\x42\xEC
\x92\x04\x43\x4E\x54\x00\x30
\x8A\x05\x10\x00\x60
\x92\x80\x03\x00\x03\x00\xED

\xA1\x00\x4C\x50\x46\x32\x2D\x54\x49\x4C\x54\x00\x00\x00\x00\x00\x00\x00\x1E <- mode 1

\x99\x01\x00\x00\x00\x00\x00\x00\x20\x41\x06
\x99\x02\x00\x00\x00\x00\x00\x00\xC8\x42\xEE
\x99\x03\x00\x00\x00\x00\x00\x00\x20\x41\x04
\x91\x04\x44\x49\x52\x00\x35
\x89\x05\x04\x00\x77
\x91\x80\x01\x00\x02\x00\xED

\xA0\x00\x4C\x50\x46\x32\x2D\x41\x4E\x47\x4C\x45\x00\x00\x00\x00\x00\x00\x5B <- mode 0

\x98\x01\x00\x00\x34\xC2\x00\x00\x34\x42\xE6
\x98\x02\x00\x00\xC8\xC2\x00\x00\xC8\x42\xE5
\x98\x03\x00\x00\x34\xC2\x00\x00\x34\x42\xE4
\x90\x04\x44\x45\x47\x00\x2D
\x88\x05\x10\x00\x62
\x90\x80\x02\x00\x03\x00\xEE

\x04 <- BYTE_ACK

\xF0 <- I have no idea what this is

\x00 <- BYTE_SYNC

\x40\x22\x9D
\x49\x03\x02\xB7
\x52\x00\xC2\x01\x00\x6E
\x5F\x00\x00\x00\x10\x00\x00\x00\x10\xA0

\x9B\x00\x4C\x50\x46\x32\x2D\x43\x41\x4C\x6F
\x9B\x01\x00\x00\x34\xC2\x00\x00\x34\x42\xE5
\x9B\x02\x00\x00\xC8\xC2\x00\x00\xC8\x42\xE6
\x9B\x03\x00\x00\x34\xC2\x00\x00\x34\x42\xE7
\x93\x04\x43\x41\x4C\x00\x26
\x8B\x05\x10\x00\x61
\x93\x80\x03\x00\x03\x00\xEC

\xA2\x00\x4C\x50\x46\x32\x2D\x43\x52\x41\x53\x48\x00\x00\x00\x00\x00\x00\x53 <- mode 2

\x9A\x01\x00\x00\x00\x00\x00\x00\xC8\x42\xEE
\x9A\x02\x00\x00\x00\x00\x00\x00\xC8\x42\xED
\x9A\x03\x00\x00\x00\x00\x00\x00\xC8\x42\xEC
\x92\x04\x43\x4E\x54\x00\x30
\x8A\x05\x10\x00\x60
\x92\x80\x03\x00\x03\x00\xED

\xA1\x00\x4C\x50\x46\x32\x2D\x54\x49\x4C\x54\x00\x00\x00\x00\x00\x00\x00\x1E <- mode 1

\x99\x01\x00\x00\x00\x00\x00\x00\x20\x41\x06
\x99\x02\x00\x00\x00\x00\x00\x00\xC8\x42\xEE
\x99\x03\x00\x00\x00\x00\x00\x00\x20\x41\x04
\x91\x04\x44\x49\x52\x00\x35
\x89\x05\x04\x00\x77
\x91\x80\x01\x00\x02\x00\xED

\xA0\x00\x4C\x50\x46\x32\x2D\x41\x4E\x47\x4C\x45\x00\x00\x00\x00\x00\x00\x5B <- mode 0

\x98\x01\x00\x00\x34\xC2\x00\x00\x34\x42\xE6
\x98\x02\x00\x00\xC8\xC2\x00\x00\xC8\x42\xE5
\x98\x03\x00\x00\x34\xC2\x00\x00\x34\x42\xE4
\x90\x04\x44\x45\x47\x00\x2D
\x88\x05\x10\x00\x62
\x90\x80\x02\x00\x03\x00\xEE

\x04  <- BYTE_ACK
````
I calculated the checksum for almost any row and they check out.

You can see that the init sequence is repeated twice, but there is a \xF0 between them, which I have no idea what it is.
Also, compared with [https://github.com/ahmedjouirou/legopup_arduino](https://github.com/ahmedjouirou/legopup_arduino), the first SYNC byte (\x00) is missing both in `ColorDistanceSensor.cpp` and `TiltSensor.cpp`

Unfortunately I don't have all the hardware to test, could you confirm that it works even without the first SYNC byte?
It would also be interesting to try and send the sequence only once and check if it is still recognized.

Anyways, I got the info on the protocol here [https://github.com/pybricks/technical-info/blob/master/uart-protocol.md](https://github.com/pybricks/technical-info/blob/master/uart-protocol.md) .

**Comment by ysard** [comment]
Hi, thanks for your comment. You did the dump formatting that I didn't have the courage to do :p

> Could you confirm that it works even without the first SYNC byte?

I can answer this question quickly.
I judged at the time that this byte was an artifact due to the logic analyzer used when listening.

It prevented me from managing the debugging with the Python library properly, it is not used in the PyBricks test sequences (see https://github.com/pybricks/pybricks-micropython/blob/11d92d05ab69b9c37b3b5b0035378d1722cbb492/lib/pbio/test/src/uartdev.c#L116), and **it is useless in real situation of testing the distance and color sensor** with the Lego hardware.

It is not justified in the Pybricks protocol documentation, you will notice unless I am mistaken that there is no mention of a SYNC byte at the beginning of the initialization sequence, but just at the end. 

Cf:
> The sequence sent by the device is terminated by BYTE_ACK and BYTE_SYNC.

You will also notice that this SYNC byte is not present after the ACK at the end of the initialization sequence.
This is more mysterious. I don't know if the PyBricks code expects to receive this byte.
In any case **the hub doesn't need it to return the NACK in response**.

So I removed the first 0x00 early during development in this commit (I should have put more info in the message): 66601e5.

> You can see that the init sequence is repeated twice, but there is a \xF0 between them

Indeed, the sequence is repeated twice. As far as I know and remember (it's already been many months, time is running fast...), this is not justified in the protocol.

Maybe it is the same as for the byte 0x00. At the moment I don't have the time to test this, although in practice it should be enough to compile the sketch given in the examples.

In my opinion the first sequence is received by the hub which returns a NACK. This NACK is put in the receiver queue of the sensor while the sensor sends the second sequence. 
Either the hub considers that its first NACK did not get a response in time and takes into account the second sequence. Or the second sequence is sent on time and only reconfigures the hub for nothing.
